2016 eoy fundraising snippet.

ScottDowne opened the pull request:
@glogiotatidis Mind taking a look at this?

ScottDowne commented:
Here is the link again about what it's doing: https://github.com/MozillaFoundation/Advocacy/issues/326

There is a little table in there with some of the tests we want to run with it.

glogiotatidis commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,321 @@
+ }
+{% endif %}
+
+{% if button_color == 'green' %}
```
Let's make this an `else if` to keep things tidy

glogiotatidis replied:
then then an `else` with the default value of blue.

ScottDowne replied:
Done.

glogiotatidis commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,321 @@
+          if (donationURL.indexOf('?') >= 0) {
+            delimiter = '&';
+          }
+          var dAmountFirst = '{{donation_amount_first}}';
```
Add spaces `{{ donation_amount_first }}`. same for next lines

glogiotatidis commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,321 @@
+  @blockable: checkbox - Check to allow user to block this snippet.
+  @block_button_text - small text - Text to display while hovering over opt out button. Default 'Remove this'
+
+#}
```
👍 Good docs, thanks!

ScottDowne replied:
Thanks!

glogiotatidis commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,321 @@
+          var btnAmounts = snippet.querySelectorAll('.donation-amount');
+          var btnSelected;
+          var amount;
+          var frequency = 'single';
```
Do we need this?

ScottDowne replied:
Nope, we do not. Good catch.

glogiotatidis commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,321 @@
+          function updateLink() {
+            btnDonate.attributes['href'].value = donationURL + delimiter + 'amount=' + amount + "&frequency=" + frequency + "&currency=" + currencyCode + '&presets=' + presets;
+          }
+          var selected = "{{ selected_button|default("donation-amount-second", true) }}";
```
Replace outer `"` with `'`

glogiotatidis replied:
Or the other way around. You mix a lot `'` with `"`, please stick to one code style :)

ScottDowne replied:
Fixed up. I went with double quotes for most everything, cept for any quotes inside the default calls in nunjucks, where I use single while inside a double quote string, they are like this: `"{{ currency_code|default('usd', true) }}"` this is to keep the syntax highlighters from getting confused.

glogiotatidis replied:
👍

glogiotatidis commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,321 @@
+          var frequency = 'single';
+
+          var locale = '{{ locale|default("en-US", true) }}';
+          var currencyCode = '{{ currency_code|default('usd', true) }}';
```
Quotes

glogiotatidis commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,321 @@
+            updateLink();
+          }
+          function updateLink() {
+            btnDonate.attributes['href'].value = donationURL + delimiter + 'amount=' + amount + "&frequency=" + frequency + "&currency=" + currencyCode + '&presets=' + presets;
```
Quotes

ScottDowne commented:
Thanks, working on these now.

ScottDowne commented:
@glogiotatidis Fixed up. I also got the block snippet button back up. It broke because I added a unique snippet id, and not the generic one, but I didn't change the line that the button used to set visibility while hovering, so it would never show up while hovering.

ScottDowne commented:
Also, I figure we can squash any commits into one before we merge, but for now, I'm leaving them separate so you can see what I changed.

glogiotatidis commented:
> @glogiotatidis Fixed up. I also got the block snippet button back up. It broke because I added a unique snippet id, and not the generic one, but I didn't change the line that the button used to set visibility while hovering, so it would never show up while hovering.

Nice. I see that `snippet_id` is used elsewhere too. Can you please remove to keep things tidy?

glogiotatidis commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,328 @@
+  @currency_code: Default usd. The code for the currency. Examle gbp, cad, usd.
+
+  @small_buttons: checkbock - Scales the buttons down.
+  @donation_amount_first: number. Required. The amount for the first button.
```
Here you say it's required and later it looks it can be skipped? https://github.com/mozmar/snippets/pull/88/files#diff-90c0ff362679dfe399e92a5a6fc3182fR217

ScottDowne replied:
Good catch here too. Fixed.

Question though, the button amounts are only optional as a group. It's all or none. I wasn't sure if I needed to explain that in the docs or not, but I can if you feel it's necessary.

glogiotatidis commented:
Contents are touching the ends of block-overlay. On simple snippet template there's more space. This is by design?

![image](https://cloud.githubusercontent.com/assets/584352/18829772/81aa4fb4-83e6-11e6-91d1-0eacd827be83.png)

![image](https://cloud.githubusercontent.com/assets/584352/18829784/901b607e-83e6-11e6-8b68-59d952a696ad.png)

ScottDowne commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,328 @@
+            amount = element.attributes["data-amount"].value;
+            updateLink();
+          }
+          function updateLink() {
```
Nope, good catch. It used to be called twice, but I simplified it a bit, and now only needed to be called once, and didn't think to just remove the function.

ScottDowne commented:
I removed the snippet-id bits. I trust you that it's safe :) but I don't actually know how things changed which makes it safe, and how much I can change given there are other things I used to do to protect the snippet from leaks, like put everything inside a `div` and I'm still using a id `mofo-eoy-fundraiser` for everything. Can I update this more?

I added some padding around the block overlay, and also tweaked it a bit to remove some js, and instead used sibling selectors and hover states. This is much cleaner and removed lots of js. I also think there is room for improvement here, but for now, I'm, not going to do enhancement stuff, and may do that later after some discussion.

glogiotatidis commented:
> I removed the snippet-id bits. I trust you that it's safe :) but I don't actually know how things changed which makes it safe, and how much I can change given there are other things I used to do to protect the snippet from leaks, like put everything inside a div and I'm still using a id mofo-eoy-fundraiser for everything. Can I update this more?

Back in the day we use to put all snippets in the DOM of about:home and hide them all except one with `display:none`. Thus we needed to uniquely name everything to avoid breaking things. Now are packing snippets in JSON and only inserting into the DOM one snippet, thus no need to uniquely name things anymore. 

Yes you can update this as much as you want. The service doesn't care how you name things.

ScottDowne commented:
@glogiotatidis I couldn't figure out how to put something back into review using githubs new review flow, but this is ready again.

ScottDowne commented:
@glogiotatidis OK cool, so I moved the snippet out of a global `div` because it's cleaner, and not necessary. I also did a bunch of whitespace cleanup as it was getting out of hand.

I left an id on the container div that's eoy related, as I did find it slightly helpful and clear, although not needed for namespacing reasons anymore.

I did this all in a new commit, and still haven't squashed anything.

glogiotatidis commented:
Re: optional buttons

Let's add a checkbox "show_donation_amount_buttons"  and check against that. This will make it more clear to snippet editors (who don't read this documentation) how to configure the snippet.

ScottDowne commented:
@glogiotatidis good idea. I updated the docs and added a show_donation_amount_buttons

If everything checks out, feel free to rebase and squash all the commits into the first. I left separate again just so you can easily see what was changed.

Thanks!

glogiotatidis commented:
Close button overlays donate button. Maybe shrink the first?

![image](https://cloud.githubusercontent.com/assets/584352/19038014/20fccbd2-8982-11e6-9e31-a12d9e3449d7.png)

ScottDowne commented:
@glogiotatidis what if I moved it to the corner, to be more like this? http://people.mozilla.org/%7Eshorlander/about-home-snippet-opt-out/about-home-test-02.html

I'm just not sure shrinking it is going to be enough.

glogiotatidis commented:
> @glogiotatidis what if I moved it to the corner, to be more like this? http://people.mozilla.org/%7Eshorlander/about-home-snippet-opt-out/about-home-test-02.html
> 
> I'm just not sure shrinking it is going to be enough.

I'd prefer to keep a unified experience as much as possible. I'd shrink and maybe push the button down.

ScottDowne commented:
k, can do.

ScottDowne commented:
@glogiotatidis so moving the button down is a bit of an issue, with the centering and such. I couldn't figure it out without breaking the centering in some way.

So I'm trying something a bit different, but might be better. I added some space in the overlay, to fix the button, which essentially pushed everything down, (so it stayed all centered) then I added a slight negative top margin to offset the change.

![snippet](https://cloud.githubusercontent.com/assets/197334/19039712/0a890d3a-894f-11e6-90f7-c377dfc6a6b9.png)

 Thoughts?

glogiotatidis commented:
if @jkt230 is ok with the latest screenshot, I'm ok

jkt230 commented:
Latest screenshot looks good. Thanks!

glogiotatidis commented on campaigns/mofo-eoy-2016/snippet.html:
```diff
@@ -0,0 +1,300 @@
+  @button_label: Required. Text for donate button.
+  @donation_form_url: Required. Url to the donation form.
+
+  @button_color: Default blue. Red and green also options.
```
Note: "red" is a bad option for button color because it's disruptive and also means a dangerous action. I know we default to blue so I won't block on this but please communicate this to folks asking for red buttons. 

It also made ppl upset in the past!

glogiotatidis commented:
Merged in 0642582a9672e75659db298396d01c35789968d2

Thanks @ScottDowne 🍻
